### PR TITLE
PARQUET-2117: Expose Row Index via ParquetReader and ParquetRecordReader

### PR DESCRIPTION
**Pull request by prakharjain09**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/PARQUET-2117
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests:
Added TestParquetReader which covers rowIndex related tests for different kind of filters.
Also Extended all the ColumnIndexFiltering and BloomFiltering tests to validate the "row index" also. This adds unit test coverage for following scenarios for this feature: Parquet V1/V2 with encryption on/off with no-filter/simple-filter/column-index-filter/bloom-filter. 

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-column/src/main/java/org/apache/parquet/column/page/PageReadStore.java
Comment (prakharjain09) [inline parquet-column/src/main/java/org/apache/parquet/column/page/PageReadStore.java:50]: @shangxinli I added this new method in the interface. This leads to build issues from japicmp-maven-plugin - METHOD_ADDED_TO_INTERFACE.

I provided a default value for this method and this seems to solve the issue. Please provide your feedback on this approach/any alternate approach?

**Comment by prakharjain09** [comment]
@shangxinli @gszadovszky Please review the changes when you get chance. Thanks!

**Comment by shangxinli** [comment]
Can you squash the commits to make the review easier?

**Comment by prakharjain09** [comment]
> [](https://github.com/prakharjain09)Can you squash the commits to make the review easier?

done

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:75]: The name is so long

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:280]: What is the reason not turning -1?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:294]: The name is so long

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:1420]: use final

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java
Comment (shangxinli) [inline parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java:343]: Removing public method could cause incompatibility issue.

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:1420]: done.

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:1435]: The `visit(OffsetMetadataFilter filter)` and `visit(RangeMetadataFilter filter)` returns the "filtered" FileMetadata and so few of the rowGroups might be missing in that. So doing it at the end will generate incorrect rowIndexOffset.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java]: Makes sense - retaining the older constructor.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java]: done.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: renamed to `rowIdxInFileItr`.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: renamed to shorter name.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:280]: current is an existing variable which tracks number of rows already processed. It is initialized to 0 at declaration time. So here we are trying to see if it is still 0, that means we haven't processed any row yet.

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java
Comment (prakharjain09) [reply parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java:343]: fixed - not deleting it.

File: parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java
Comment (prakharjain09) [reply parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java:320]: This is being used by the new test file - https://github.com/apache/parquet-mr/pull/945/files#diff-276ac02899424a4245b8589f8ee6d444e3619f6be17834e4d5d7e81dfdaaee39R129

**Comment by prakharjain09** [comment]
@shangxinli Thanks fa lot or the review.

> [](https://github.com/prakharjain09)We need more test to cover old parquet data that doesn't have column index.

I couldn't find any existing tests or any existing parquet files in Resource directory which doesn't have column indexes. Could you please give some pointer to similar existing test or some way to create parquet file without column indexes (don't see any options to disable writing column indexes either)?
I have added tests for validating row indexes are correct with "column index filtering" disabled.

**Comment by shangxinli** [comment]
I will have another look soon sometime this week.

**Comment by prakharjain09** [comment]
@shangxinli I have added test to cover old parquet file without column indexes. Please review the changes when you get chance.

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (shangxinli) [reply parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:1435]: Got it.

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:1512]: As you mentioned above, if parquetMetadata is a filtered one, then generateRowGroupOffsets() won't return accurate offsets, correct?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/BlockMetaData.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/BlockMetaData.java:36]: In the following toString(), it should be added too.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:235]: && hasNext()?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:301]: The name of 'column index' was already used for Page Index in another feature. Can you use something else?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (shangxinli) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:280]: I understand why we are checking 'current == 0L'. I was asking why you choose throw exception other than returning an invalid value. This is a public method. We should have it documented ether way you choose.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: done.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/BlockMetaData.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/BlockMetaData.java]: done.

File: parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java]: Yes thats correct. Fixed this - now we are passing empty Map so that we don't populate incorrect rowIndexOffsets.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: removed this code comment.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: There was no specific reason for choosing exception over -1.
I have updated it to return -1 and also updated all the public method docs to reflect the same.

**Comment by prakharjain09** [comment]
@shangxinli Thanks a lot or the review. I have addressed the review comments. Could you please look into the PR again?

Also could you share information about when are we planning to do code-freeze for next minor release? It will be great if we can release this change in next minor/patch release so that Apache Spark/other projects get to use this functionality sooner.

**Comment by prakharjain09** [comment]
> Could you please look into the PR again?
> Also could you share information about when are we planning to do code-freeze for next minor release? It will be great if we can release this change in next minor/patch release so that Apache Spark/other projects get to use this functionality sooner.


@shangxinli Gentle reminder. Please take a look when you get chance.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetReader.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetReader.java:144]: Given this is public method, we need to take care of the Java doc decrations

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetReader.java
Comment (shangxinli) [inline parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetReader.java:50]: I am not sure if it is a good idea to check in a data file. Can you check if it is possible to stop generating offset index in the current version of Parquet?

**Comment by shangxinli** [comment]
I just left some comments. Other than that, it looks good to me. Add @ggershinsky in case you have time to have a look.

File: parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetReader.java:50]: @shangxinli It looks like the [column-indexes](https://issues.apache.org/jira/browse/PARQUET-1201) are always written in the current version of parquet and are not configurable.
We are already testing the new row index support with and without the column index filtering being triggered (as part of TestColumnIndexFiltering). Also the new row index feature doesn't rely on column indexes in any way. So we can skip the backward compatibility testing and remove this parquet file from resources. What do you think about this?

**Comment by ggershinsky** [comment]
hi guys, I'm OOO (vacation) this week. Can review it next week if helps, but feel free to go ahead without waiting for me.

**Comment by prakharjain09** [comment]
@shangxinli Thanks for taking another look. I have addressed all comments other [than one](https://github.com/apache/parquet-mr/pull/945#discussion_r820928524). Please advice on the same. Thanks!

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (ggershinsky) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:316]: nit: could you start the method with checking this condition (!rowGroupRowIdxOffset.isPresent()), and then return? Will look cleaner.

**Comment by ggershinsky** [comment]
thanks for this change. The PR looks good to me now, I'll add my approval after it passes the CI tests.

**Comment by shangxinli** [comment]
@prakharjain09 After you fix the CI failures, we can merge.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (prakharjain09) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: done.

**Comment by prakharjain09** [comment]
Thanks @ggershinsky for the review. I have addressed the comments and fixed the build issue.

**Review by shangxinli — APPROVED** [review]

**Comment by prakharjain09** [comment]
@shangxinli @ggershinsky Thanks a lot for reviewing this change.

This will unblock SPARK-37980 if this is released as part of upcoming parquet release. Do we need to cherry-pick this to any release branch for the same?

**Comment by ggershinsky** [comment]
@prakharjain09 the upcoming parquet release will include the current master (plus a couple of WIP PRs, once they are merged), so this patch will be covered.

**Comment by ggershinsky** [comment]
@prakharjain09 hopefully, we'll resolve the remaining issues at the community sync tomorrow, and start working on a cut.

**Pull request merged**
